### PR TITLE
Projection def stepping

### PR DESCRIPTION
**Pull request by jburel**
# What this PR does

Add a new column to projection def

# Testing this PR

Note that this PR introduces a new DB patch 9.

Ensure that there are no regressions related to rendering settings.

# Related reading

https://trello.com/c/JkRcxqVU/96-z-projection-thumbnails

### CODE REVIEW

**Comment by will-moore** [comment]
Model and DB changes look good (to my untrained eye). Are we actually using any of the Projected defs yet? Projection in Insight works OK, but we're not saving projections yet, right?

**Comment by will-moore** [comment]
Ah, I realise now it's on "breaking"!

**Comment by jburel** [comment]
We are not using projectionDef at all yet. 
Only in the test. 
I have started working on a branch introducing the "axis" I hope to have it in the breaking queue next week.

**Comment by aleksandra-tarkowska** [comment]
This PR broke daily build this night. For some reason omego couldn't upgrade DB properly

https://ci.openmicroscopy.org/job/OMERO-DEV-merge-deploy/384/console

```
***************************************************************************************
DB version (OMERO5.3DEV__8) does not match omero.properties (OMERO5.3DEV__9). Please apply a db upgrade.
See http://www.openmicroscopy.org/site/support/omero5.2/sysadmins/server-upgrade.html
***************************************************************************************
```

rerunning job https://ci.openmicroscopy.org/job/OMERO-DEV-merge-deploy/385/console cause 

```
2016-08-12 23:30:53,354 [omego.extern] INFO  Executing [custom environment]: psql -d OMERO-DEV-merge-deploy -h localhost -U omero -w -A -t -f /opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/sql/psql/OMERO5.3DEV__9/OMERO5.3DEV__8.sql
2016-08-12 23:30:53,374 [omego.extern] INFO  Completed [0.019 s]
2016-08-12 23:30:53,374 [    omego.db] WARNI stderr: psql:/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/sql/psql/OMERO5.3DEV__9/OMERO5.3DEV__8.sql:54: ERROR:  column "stepping" of relation "projectiondef" already exists
psql:/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/sql/psql/OMERO5.3DEV__9/OMERO5.3DEV__8.sql:64: ERROR:  current transaction is aborted, commands ignored until end of transaction block
psql:/opt/hudson/workspace/OMERO-DEV-merge-deploy/src/dist/sql/psql/OMERO5.3DEV__9/OMERO5.3DEV__8.sql:66: ERROR:  current transaction is aborted, commands ignored until end of transaction block
```

**Comment by jburel** [comment]
Moving it back to breaking and I will retest.

**Comment by jburel** [comment]
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-deploy/386/ back to green

File: components/model/resources/mappings/display.ome.xml [hunk 1]
```diff
@@ -72,6 +72,7 @@
 			<required name="active" type="boolean"/>
 			<optional name="startPlane" type="NonNegativeInteger"/>
 			<optional name="endPlane" type="NonNegativeInteger"/>
+			<optional name="stepping" type="NonNegativeInteger"/>
```
Comment (mtbc) [inline components/model/resources/mappings/display.ome.xml]: Ought this be a `PositiveInteger`?

File: components/tools/OmeroJava/test/integration/RenderingEngineTest.java [hunk 2]
```diff
@@ -1910,6 +1910,7 @@ public void testRenderingEngineProjectionDef() throws Exception {
         pDef.setActive(omero.rtypes.rbool(true));
         pDef.setStartPlane(omero.rtypes.rint(0));
         pDef.setEndPlane(omero.rtypes.rint(1));
+        pDef.setStepping(omero.rtypes.rint(1));
```
Comment (mtbc) [inline components/tools/OmeroJava/test/integration/RenderingEngineTest.java]: maybe pick a different number for each property? then can rule out property values somehow getting swapped.

**Comment by mtbc** [comment]
We might someday need a doc/comment somewhere to say what the stepping means exactly, e.g., what if `endPlane` isn't neatly `startPlane + n * stepping`.

File: sql/psql/OMERO5.3DEV__9/schema.sql [hunk 3]
```diff
@@ -1993,6 +1993,7 @@
         permissions int8 not null,
         endPlane nonnegative_int,
         startPlane nonnegative_int,
+        stepping nonnegative_int,
```
Comment (mtbc) [inline sql/psql/OMERO5.3DEV__9/schema.sql]: should be positive int

**Comment by mtbc** [comment]
This is now looking to me to be good to merge, now we just wait for CI to turn green. :+1:

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-DEV-merge-push#413](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/413/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/413/consoleText) for more details.
Possible conflicts:
  - PR #4753 jburel 'Lut rnd'
    - components/tools/OmeroJava/test/integration/RenderingEngineTest.java

**Comment by mtbc** [comment]
Given conflict: fine with me if you force-push the integration test changes from this PR and move that to a different one.

**Comment by jburel** [comment]
@mtbc: back to green. I will fix the test once https://github.com/openmicroscopy/openmicroscopy/pull/4753 is merged

**Comment by jburel** [comment]
Moving this PR back to breaking so we can do the changes related to codomain map cf. https://trello.com/c/9iAe3Sg9/95-lut-support

**Comment by mtbc** [comment]
dev merge will still deploy okay on Monday morning?

**Comment by aleksandra-tarkowska** [comment]
It won't. there is not downgrade script. @jburel could you remove the flag please?

**Comment by jburel** [comment]
ok I will work on top of this PR and opened a new one for breaking

**Comment by joshmoore** [comment]
Since any comments in the mapping files wouldn't really be consumable by devs, I'd say let's slate @mtbc's https://github.com/openmicroscopy/openmicroscopy/pull/4781#issuecomment-239771565 for inclusion in the specification documentation. Perhaps a card with the intended semantics when the two don't match?

Merging.

**Comment by mtbc** [comment]
Noted on https://trello.com/c/ghvhqiO2/34-omero-developer-docs-for-model-changes.

**Pull request merged**